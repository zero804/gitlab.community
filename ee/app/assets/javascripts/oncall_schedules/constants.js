export const LENGTH_ENUM = {
  hours: 'HOURS',
  days: 'DAYS',
  weeks: 'WEEKS',
};

export const CHEVRON_SKIPPING_SHADE_ENUM = ['500', '600', '700', '800', '900', '950'];

export const CHEVRON_SKIPPING_PALETTE_ENUM = ['blue', 'orange', 'aqua', 'green', 'magenta'];

export const HOURS_IN_DAY = 24;

export const addRotationModalId = 'addRotationModal';
export const editRotationModalId = 'editRotationModal';
