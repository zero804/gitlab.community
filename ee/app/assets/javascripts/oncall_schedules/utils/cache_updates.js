import produce from 'immer';
import createFlash from '~/flash';

import {
  DELETE_SCHEDULE_ERROR,
  UPDATE_SCHEDULE_ERROR,
  UPDATE_ROTATION_ERROR,
} from './error_messages';

const addScheduleToStore = (store, query, { oncallSchedule: schedule }, variables) => {
  if (!schedule) {
    return;
  }

  const sourceData = store.readQuery({
    query,
    variables,
  });

  const data = produce(sourceData, draftData => {
    draftData.project.incidentManagementOncallSchedules.nodes.push(schedule);
  });

  store.writeQuery({
    query,
    variables,
    data,
  });
};

const deleteScheduleFromStore = (store, query, { oncallScheduleDestroy }, variables) => {
  const schedule = oncallScheduleDestroy?.oncallSchedule;
  if (!schedule) {
    return;
  }

  const sourceData = store.readQuery({
    query,
    variables,
  });

  const data = produce(sourceData, draftData => {
    // eslint-disable-next-line no-param-reassign
    draftData.project.incidentManagementOncallSchedules.nodes = draftData.project.incidentManagementOncallSchedules.nodes.filter(
      ({ id }) => id !== schedule.id,
    );
  });

  store.writeQuery({
    query,
    variables,
    data,
  });
};

const updateScheduleFromStore = (store, query, { oncallScheduleUpdate }, variables) => {
  const schedule = oncallScheduleUpdate?.oncallSchedule;
  if (!schedule) {
    return;
  }

  const sourceData = store.readQuery({
    query,
    variables,
  });

  const data = produce(sourceData, draftData => {
    // eslint-disable-next-line no-param-reassign
    draftData.project.incidentManagementOncallSchedules.nodes = [
      ...draftData.project.incidentManagementOncallSchedules.nodes,
      schedule,
    ];
  });

  store.writeQuery({
    query,
    variables,
    data,
  });
};

const addRotationToStore = (
  store,
  query,
  { oncallRotationCreate: rotation },
  scheduleId,
  variables,
) => {
  if (!rotation) {
    return;
  }

  const sourceData = store.readQuery({
    query,
    variables,
  });

  // TODO: This needs the rotation backend to be fully integrated to work, for the moment we will place-hold it.
  const data = produce(sourceData, draftData => {
    const rotations = [rotation];

    // eslint-disable-next-line no-param-reassign
    draftData.project.incidentManagementOncallSchedules.nodes.find(
      ({ iid }) => iid === scheduleId,
    ).rotations = rotations;
  });

  store.writeQuery({
    query,
    variables,
    data,
  });
};

const updateRotationFromStore = (store, query, { oncallRotationUpdate }, scheduleId, variables) => {
  const rotation = oncallRotationUpdate?.oncallRotation;
  if (!rotation) {
    return;
  }

  const sourceData = store.readQuery({
    query,
    variables,
  });

  const data = produce(sourceData, draftData => {
    // eslint-disable-next-line no-param-reassign
    draftData.project.incidentManagementOncallSchedules.nodes = [
      ...draftData.project.incidentManagementOncallSchedules.nodes,
      rotation,
    ];
  });

  store.writeQuery({
    query,
    variables,
    data,
  });
};

const onError = (data, message) => {
  createFlash({ message });
  throw new Error(data.errors);
};

export const hasErrors = ({ errors = [] }) => errors?.length;

export const updateStoreOnScheduleCreate = (store, query, data, variables) => {
  if (!hasErrors(data)) {
    addScheduleToStore(store, query, data, variables);
  }
};

export const updateStoreAfterScheduleDelete = (store, query, data, variables) => {
  if (hasErrors(data)) {
    onError(data, DELETE_SCHEDULE_ERROR);
  } else {
    deleteScheduleFromStore(store, query, data, variables);
  }
};

export const updateStoreAfterScheduleEdit = (store, query, data, variables) => {
  if (hasErrors(data)) {
    onError(data, UPDATE_SCHEDULE_ERROR);
  } else {
    updateScheduleFromStore(store, query, data, variables);
  }
};

export const updateStoreAfterRotationAdd = (store, query, data, scheduleId, variables) => {
  if (!hasErrors(data)) {
    addRotationToStore(store, query, data, scheduleId, variables);
  }
};

export const updateStoreAfterRotationEdit = (store, query, data, scheduleId, variables) => {
  if (hasErrors(data)) {
    onError(data, UPDATE_ROTATION_ERROR);
  } else {
    updateRotationFromStore(store, query, data, scheduleId, variables);
  }
};
