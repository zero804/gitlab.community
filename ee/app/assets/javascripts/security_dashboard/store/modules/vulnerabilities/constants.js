import httpStatusCodes from '~/lib/utils/http_status';

export {
  CRITICAL,
  HIGH,
  MEDIUM,
  LOW,
  INFO,
  UNKNOWN,
  SEVERITIES,
} from '~/vulnerabilities/constants';

export const LOADING_VULNERABILITIES_ERROR_CODES = {
  UNAUTHORIZED: httpStatusCodes.UNAUTHORIZED,
  FORBIDDEN: httpStatusCodes.FORBIDDEN,
};

export const VULNERABILITIES_REQUEST_DEBOUNCE_TIME = 200;
