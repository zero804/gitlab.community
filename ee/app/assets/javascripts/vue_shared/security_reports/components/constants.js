import { s__ } from '~/locale';

export * from '~/vue_shared/security_reports/components/constants';

export const SEVERITY_TOOLTIP_TITLE_MAP = {
  unknown: s__(
    `SecurityReports|The rating "unknown" indicates that the underlying scanner doesnâ€™t contain or provide a severity rating.`,
  ),
};

export const VULNERABILITY_MODAL_ID = 'modal-mrwidget-security-issue';
