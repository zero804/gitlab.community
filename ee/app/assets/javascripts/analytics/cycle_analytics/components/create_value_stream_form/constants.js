import { __, s__ } from '~/locale';
import { capitalizeFirstCharacter } from '~/lib/utils/text_utility';

export const I18N = {
  FORM_TITLE: __('Create Value Stream'),
  FORM_CREATED: __("'%{name}' Value Stream created"),
  RECOVER_HIDDEN_STAGE: s__('CreateValueStreamForm|Recover hidden stage'),
  RESTORE_HIDDEN_STAGE: s__('CreateValueStreamForm|Restore stage'),
  RECOVER_STAGE_TITLE: s__('CreateValueStreamForm|Default stages'),
  RECOVER_STAGES_VISIBLE: s__('CreateValueStreamForm|All default stages are currently visible'),
  SELECT_START_EVENT: s__('CreateValueStreamForm|Select start event'),
  SELECT_END_EVENT: s__('CreateValueStreamForm|Select end event'),
  FORM_FIELD_NAME_LABEL: s__('CreateValueStreamForm|Name'),
  FORM_FIELD_NAME_PLACEHOLDER: s__('CreateValueStreamForm|Enter a name for the stage'),
  FIELD_STAGE_NAME_PLACEHOLDER: s__('CreateValueStreamForm|Enter stage name'),
  FORM_FIELD_START_EVENT: s__('CreateValueStreamForm|Start event'),
  FORM_FIELD_START_EVENT_LABEL: s__('CreateValueStreamForm|Start event label'),
  FORM_FIELD_END_EVENT: s__('CreateValueStreamForm|End event'),
  FORM_FIELD_END_EVENT_LABEL: s__('CreateValueStreamForm|End event label'),
  BTN_UPDATE_STAGE: s__('CreateValueStreamForm|Update stage'),
  BTN_ADD_STAGE: s__('CreateValueStreamForm|Add stage'),
  TITLE_EDIT_STAGE: s__('CreateValueStreamForm|Editing stage'),
  TITLE_ADD_STAGE: s__('CreateValueStreamForm|New stage'),
  BTN_CANCEL: __('Cancel'),
  STAGE_INDEX: s__('CreateValueStreamForm|Stage %{index}'),
  HIDDEN_DEFAULT_STAGE: s__('CreateValueStreamForm|%{name} (default)'),
};

export const ERRORS = {
  MIN_LENGTH: s__('CreateValueStreamForm|Name is required'),
  MAX_LENGTH: s__('CreateValueStreamForm|Maximum length 100 characters'),
  START_EVENT_REQUIRED: s__('CreateValueStreamForm|Please select a start event first'),
  STAGE_NAME_EXISTS: s__('CreateValueStreamForm|Stage name already exists'),
  INVALID_EVENT_PAIRS: s__(
    'CreateValueStreamForm|Start event changed, please select a valid end event',
  ),
};

export const STAGE_SORT_DIRECTION = {
  UP: 'UP',
  DOWN: 'DOWN',
};

export const NAME_MAX_LENGTH = 100;

export const defaultErrors = {
  id: [],
  name: [],
  startEventIdentifier: [],
  startEventLabelId: [],
  endEventIdentifier: [],
  endEventLabelId: [],
};

export const defaultFields = {
  id: null,
  name: null,
  startEventIdentifier: null,
  startEventLabelId: null,
  endEventIdentifier: null,
  endEventLabelId: null,
};

export const DEFAULT_STAGE_CONFIG = ['issue', 'plan', 'code', 'test', 'review', 'staging'].map(
  (id, i) => ({
    id,
    defaultName: capitalizeFirstCharacter(id),
    name: capitalizeFirstCharacter(id),
    startEventIdentifier: i % 3 === 0 ? __('Label added') : __('Issue Created'),
    endEventIdentifier: i % 3 === 0 ? __('Label removed') : __('Issue closed'),
    startEventLabel: i % 3 === 0 ? __('Label a') : null,
    endEventLabel: i % 3 === 0 ? __('Label b') : null,
    custom: false,
    hidden: false,
  }),
);
