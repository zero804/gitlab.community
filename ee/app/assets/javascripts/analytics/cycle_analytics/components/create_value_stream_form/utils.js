import { isStartEvent, getAllowedEndEvents, eventToOption, eventsByIdentifier } from '../../utils';
import { I18N, ERRORS, defaultErrors, defaultFields } from './constants';
import { DEFAULT_STAGE_NAMES } from '../../constants';

/**
 * @typedef {Object} CustomStageEvents
 * @property {String} canBeStartEvent - Title of the metric measured
 * @property {String} name - Friendly name for the event
 * @property {String} identifier - snakeized name for the event
 *
 * @typedef {Object} DropdownData
 * @property {String} text - Friendly name for the event
 * @property {String} value - Value to be submitted for the dropdown
 */

/**
 * Takes an array of custom stage events to return only the
 * events where `canBeStartEvent` is true and converts them
 * to { value, text } pairs for use in dropdowns
 *
 * @param {CustomStageEvents[]} events
 * @returns {DropdownData[]} array of start events formatted for dropdowns
 */
export const startEventOptions = eventsList => [
  { value: null, text: I18N.SELECT_START_EVENT },
  ...eventsList.filter(isStartEvent).map(eventToOption),
];

/**
 * Takes an array of custom stage events to return only the
 * events where `canBeStartEvent` is false and converts them
 * to { value, text } pairs for use in dropdowns
 *
 * @param {CustomStageEvents[]} events
 * @returns {DropdownData[]} array end events formatted for dropdowns
 */
export const endEventOptions = (eventsList, startEventIdentifier) => {
  const endEvents = getAllowedEndEvents(eventsList, startEventIdentifier);
  return [
    { value: null, text: I18N.SELECT_END_EVENT },
    ...eventsByIdentifier(eventsList, endEvents).map(eventToOption),
  ];
};

/**
 * @typedef {Object} CustomStageFormData
 * @property {Object.<String, String>} fields - form field values
 * @property {Object.<String, Array>} errors - form field errors
 */

/**
 * Initializes the fields and errors for the custom stages form
 * providing defaults for any missing keys
 *
 * @param {CustomStageFormData} data
 * @returns {CustomStageFormData} the updated initial data with all defaults
 */
export const initializeFormData = ({ fields, errors }) => {
  const initErrors = fields?.endEventIdentifier
    ? defaultErrors
    : {
        ...defaultErrors,
        endEventIdentifier: !fields?.startEventIdentifier ? [ERRORS.START_EVENT_REQUIRED] : [],
      };
  return {
    fields: {
      ...defaultFields,
      ...fields,
    },
    errors: {
      ...initErrors,
      ...errors,
    },
  };
};

/**
 * Validates the form fields for the custom stages form
 * Any errors will be returned in a object where the key is
 * the name of the field.g
 *
 * @param {Object} fields key value pair of form field values
 * @returns {Object} key value pair of form fields with an array of errors
 */
export const validateFields = fields => {
  const newErrors = {};

  if (fields?.name) {
    newErrors.name = DEFAULT_STAGE_NAMES.includes(fields?.name.toLowerCase())
      ? [ERRORS.STAGE_NAME_EXISTS]
      : [];
  }

  if (fields?.startEventIdentifier) {
    newErrors.endEventIdentifier = [];
  } else {
    newErrors.endEventIdentifier = [ERRORS.START_EVENT_REQUIRED];
  }

  if (fields?.startEventIdentifier && fields?.endEventIdentifier) {
    newErrors.endEventIdentifier = [];
  }
  return newErrors;
};
