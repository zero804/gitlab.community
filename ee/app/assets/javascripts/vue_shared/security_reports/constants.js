/* eslint-disable import/export */
import { invert } from 'lodash';
import { reportTypeToSecurityReportTypeEnum as reportTypeToSecurityReportTypeEnumCE } from '~/vue_shared/security_reports/constants';

export * from '~/vue_shared/security_reports/constants';

/**
 * Security scan report types, as provided by the backend.
 */
export const REPORT_TYPE_API_FUZZING = 'api_fuzzing';

/**
 * SecurityReportTypeEnum values for use with GraphQL.
 *
 * These should correspond to the lowercase security scan report types.
 */
export const SECURITY_REPORT_TYPE_ENUM_API_FUZZING = 'API_FUZZING';

/* Override CE Definitions */

/**
 * A mapping from security scan report types to SecurityReportTypeEnum values.
 */
export const reportTypeToSecurityReportTypeEnum = {
  ...reportTypeToSecurityReportTypeEnumCE,
  [REPORT_TYPE_API_FUZZING]: SECURITY_REPORT_TYPE_ENUM_API_FUZZING,
};

/**
 * A mapping from SecurityReportTypeEnum values to security scan report types.
 */
export const securityReportTypeEnumToReportType = invert(reportTypeToSecurityReportTypeEnum);
