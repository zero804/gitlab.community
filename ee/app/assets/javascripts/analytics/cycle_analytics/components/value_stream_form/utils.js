import { isStartEvent, getAllowedEndEvents, eventToOption, eventsByIdentifier } from '../../utils';
import { s__ } from '~/locale';

const I18N = {
  SELECT_START_EVENT: s__('CustomCycleAnalytics|Select start event'),
  SELECT_END_EVENT: s__('CustomCycleAnalytics|Select stop event'),
};

export const startEventOptions = eventsList => [
  { value: null, text: I18N.SELECT_START_EVENT },
  ...eventsList.filter(isStartEvent).map(eventToOption),
];

export const endEventOptions = (eventsList, startEventIdentifier) => {
  const endEvents = getAllowedEndEvents(eventsList, startEventIdentifier);
  return [
    { value: null, text: I18N.SELECT_END_EVENT },
    ...eventsByIdentifier(eventsList, endEvents).map(eventToOption),
  ];
};
