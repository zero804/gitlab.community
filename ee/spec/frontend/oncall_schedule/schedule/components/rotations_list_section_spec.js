import { shallowMount } from '@vue/test-utils';
import { GlCard } from '@gitlab/ui';
import RotationsListSection from 'ee/oncall_schedules/components/schedule/components/rotations_list_section.vue';
import CurrentDayIndicator from 'ee/oncall_schedules/components/schedule/components/current_day_indicator.vue';
import { getTimeframeForWeeksView } from 'ee/oncall_schedules/components/schedule/utils';
import { PRESET_TYPES } from 'ee/oncall_schedules/components/schedule/constants';
import mockRotations from '../../mocks/mock_rotation.json';

describe('RotationsListSectionComponent', () => {
  let wrapper;
  const mockTimeframeInitialDate = new Date(2018, 0, 1);
  const mockTimeframeWeeks = getTimeframeForWeeksView(mockTimeframeInitialDate);

  function mountComponent({
    presetType = PRESET_TYPES.WEEKS,
    timeframe = mockTimeframeWeeks,
  } = {}) {
    wrapper = shallowMount(RotationsListSection, {
      propsData: {
        presetType,
        timeframe,
        rotations: mockRotations,
      },
      stubs: {
        GlCard,
      },
    });
  }

  beforeEach(() => {
    mountComponent();
  });

  afterEach(() => {
    if (wrapper) {
      wrapper.destroy();
      wrapper = null;
    }
  });

  const findTimelineCells = () => wrapper.findAll('[data-testid="timelineCell"]');

  it('renders component layout', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders timeline cell items based on timeframe data', () => {
    expect(findTimelineCells().length).toBe(mockTimeframeWeeks.length * 2);
  });

  it('renders current day indicator in the first timeline cell', () => {
    expect(
      findTimelineCells()
        .at(0)
        .find(CurrentDayIndicator)
        .exists(),
    ).toBe(true);
  });
});
