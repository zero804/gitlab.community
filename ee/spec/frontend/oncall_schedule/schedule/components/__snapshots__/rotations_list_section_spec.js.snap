// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotationsListSectionComponent renders component layout 1`] = `
<div
  class="list-section"
>
  <div
    class="list-item list-item-empty clearfix"
  >
    <span
      class="details-cell gl-display-flex gl-justify-content-space-between gl-align-items-center gl-pl-3"
    >
      <span
        class="gl-str-truncated"
      >
        Rotation 242
      </span>
       
      <gl-button-group-stub
        class="gl-px-2"
      >
        <gl-button-stub
          aria-label="Edit rotation"
          buttontextclasses=""
          category="tertiary"
          icon="pencil"
          role="button"
          size="medium"
          tabindex="0"
          title="Edit rotation"
          variant="default"
        />
         
        <gl-button-stub
          aria-label="Delete rotation"
          buttontextclasses=""
          category="tertiary"
          icon="remove"
          role="button"
          size="medium"
          tabindex="0"
          title="Delete rotation"
          variant="default"
        />
      </gl-button-group-stub>
    </span>
     
    <span
      class="timeline-cell"
      data-testid="timelineCell"
    >
      <current-day-indicator-stub
        presettype="WEEKS"
        timeframeitem="Mon Jan 01 2018 00:00:00 GMT+0000 (Greenwich Mean Time)"
      />
    </span>
    <span
      class="timeline-cell"
      data-testid="timelineCell"
    >
      <current-day-indicator-stub
        presettype="WEEKS"
        timeframeitem="Mon Jan 08 2018 00:00:00 GMT+0000 (Greenwich Mean Time)"
      />
    </span>
  </div>
  <div
    class="list-item list-item-empty clearfix"
  >
    <span
      class="details-cell gl-display-flex gl-justify-content-space-between gl-align-items-center gl-pl-3"
    >
      <span
        class="gl-str-truncated"
      >
        Rotation 244
      </span>
       
      <gl-button-group-stub
        class="gl-px-2"
      >
        <gl-button-stub
          aria-label="Edit rotation"
          buttontextclasses=""
          category="tertiary"
          icon="pencil"
          role="button"
          size="medium"
          tabindex="0"
          title="Edit rotation"
          variant="default"
        />
         
        <gl-button-stub
          aria-label="Delete rotation"
          buttontextclasses=""
          category="tertiary"
          icon="remove"
          role="button"
          size="medium"
          tabindex="0"
          title="Delete rotation"
          variant="default"
        />
      </gl-button-group-stub>
    </span>
     
    <span
      class="timeline-cell"
      data-testid="timelineCell"
    >
      <current-day-indicator-stub
        presettype="WEEKS"
        timeframeitem="Mon Jan 01 2018 00:00:00 GMT+0000 (Greenwich Mean Time)"
      />
    </span>
    <span
      class="timeline-cell"
      data-testid="timelineCell"
    >
      <current-day-indicator-stub
        presettype="WEEKS"
        timeframeitem="Mon Jan 08 2018 00:00:00 GMT+0000 (Greenwich Mean Time)"
      />
    </span>
  </div>
</div>
`;
